text/0112-server-runtime.md

```diff
@@ -0,0 +1,331 @@
+---
+title: LWC server runtime
+status: DRAFTED
+created_at: 2020-01-06
+updated_at: 2020-01-06
+pr: (leave this empty until the PR is created)
+---
+
+# LWC server runtime
+
+## Summary
+
+The LWC engine has been designed from the beginning to run in an environment with access to the DOM APIs. To accommodate server-side rendering (SSR) requirements, we need a way to decouple the engine from the DOM APIs. The existing `@lwc/engine` will be replaced by 3 new packages: 
+- `@lwc/engine-core` exposes platform-agnostic APIs and will be used by the different runtime packages to share the common logic.
+- `@lwc/engine-dom` exposes LWC APIs available on the browser.
+- `@lwc/engine-server` exposes LWC APIs used for server-side rendering.
+
+## Scope
+
+LWC SSR support is a broad subject, this proposal only focusses on a subset of it. This proposal covers the following topics:
+
+- How to evaluate the existing LWC engine module in a javascript context without DOM access.
+- What LWC APIs should be exposed in both environments.
+- How the LWC engine share the same logic when evaluated in both environments.
+
+The following topics are considered out of the scope of this proposal:
+
+- How the LWC component tree rendered on the server gets serialized and sent to the browser.
+- How the serialized component tree gets rehydrated when processed by the browser.
+
+## Motivation
+
+The first step in enabling SSR support on LWC is to be able to run the engine in a non-DOM-enabled environment. It is currently impossible to evaluate the engine in such environment due to its hard dependencies on DOM APIs.
+
+While we want to share as much logic as possible between the different environments and make SSR as transparent as possible for component authors, the runtime behavior and APIs exposed by the engine greatly differs between environments. In this regard, it makes sense to break up the current `@lwc/engine` package into multiple packages specifically tailored for each environment.
+
+Distributing multiple versions of the LWC engine offers the capability to expose different APIs depending on the execution context. On one hand, the current implementation of `createElement` with reaction hooks doesn't make sense in Node.js since there is no DOM tree to attach the created element. In the same way, the `renderToString` method doesn't make sense to ship in a browser.
```

Comment:
> with reaction hooks 

that's the key part, right? Since it does make sense to create an element on the server?

Reply:
The reaction hook is the key part here you are right, reaction hooks is the way we use to get notified when an element is inserted or removed from the DOM. For environments without DOM, hooks are unnecessary.